File: synclite-code-samples/synclite-logger/python/synclite_streaming.py
"""Python mirror of `synclite_streaming.rs`.

STREAMING-device sample. Same `Connection` API as the txn / store
samples; only `device-type` differs (configured in the conf file).
"""

import synclite as sl

DB_PATH = "sample_streaming_sqlite.db"
DEVICE_NAME = "sampledevice"
CONF_PATH = "sample_streaming.conf"


def write_conf() -> None:
    with open(CONF_PATH, "w", encoding="utf-8") as f:
        f.write(
            "device-name=sample-streaming\n"
            "db-engine=SQLITE\n"
            "device-type=STREAMING\n"
            f"db-path={DB_PATH}\n"
            "local-data-stage-directory=synclite-stage\n"
        )


def main() -> None:
    write_conf()

    sl.initialize(
        device_type="STREAMING",
        device_name=DEVICE_NAME,
        db_path=DB_PATH,
        config_path=CONF_PATH,
    )

    conn = sl.Connection.open_with_config(CONF_PATH)

    conn.execute(
        "CREATE TABLE IF NOT EXISTS events("
        "  ts BIGINT, event_type TEXT, payload TEXT)"
    )

    stmt = conn.prepare("INSERT INTO events(ts, event_type, payload) VALUES(?, ?, ?)")
    stmt.execute([1714200000000, "SIGNUP", '{"user":"alice"}'])
    stmt.execute([1714200001000, "LOGIN",  '{"user":"alice"}'])

    stmt = conn.prepare("INSERT INTO events(ts, event_type, payload) VALUES(?, ?, ?)")
    for i in range(10):
        stmt.add_batch([1714200002000 + i, "HEARTBEAT", '{"i":%d}' % i])
    stmt.execute_batch()

    conn.close()


if __name__ == "__main__":
    main()

File: synclite-code-samples/synclite-logger/rust/synclite_streaming.rs
//! Streaming device: append-only event ingestion. UPDATE / DELETE are
//! rejected by design — every event flows through the change log to the
//! configured destination.

use synclite::{DeviceType, Result, Value};
use synclite::rusqlite::Connection;
use synclite::SyncLiteOptions;

const DB_PATH: &str = "sample_streaming.db";
const DEVICE_NAME: &str = "sampledevice";

fn main() -> Result<()> {
    synclite::initialize(
        DeviceType::Streaming,
        DEVICE_NAME,
        DB_PATH,
        None,
        SyncLiteOptions::default(),
    )?;

    let mut conn = Connection::open(DB_PATH)?;

    conn.execute(
        "CREATE TABLE IF NOT EXISTS events(id INTEGER PRIMARY KEY, category TEXT, amount INTEGER)",
        &[],
    )?;

    {
        let mut stmt = conn.prepare("INSERT INTO events(id, category, amount) VALUES(?, ?, ?)")?;
        stmt.execute(&[Value::Int(1), Value::Text("stream".into()), Value::Int(40)])?;
        stmt.execute(&[Value::Int(2), Value::Text("stream".into()), Value::Int(60)])?;
    }

    // STREAMING devices accept INSERTs and DDL but reject UPDATE / DELETE.
    let update_err = conn
        .execute(
            "UPDATE events SET amount = ? WHERE id = ?",
            &[Value::Int(90), Value::Int(2)],
        )
        .expect_err("streaming should reject UPDATE");
    println!("UPDATE rejected: {update_err}");

    let delete_err = conn
        .execute("DELETE FROM events WHERE id = ?", &[Value::Int(1)])
        .expect_err("streaming should reject DELETE");
    println!("DELETE rejected: {delete_err}");

    conn.commit()?;

    conn.flush()?;
    synclite::await_sync(DB_PATH, std::time::Duration::from_secs(30))?;
    conn.close()?;
    Ok(())
}
